gradio/app.py

```diff
@@ -0,0 +1,314 @@
+"""Implements a FastAPI server to run the gradio interface."""
+
+from __future__ import annotations
+from fastapi import FastAPI, Form, Request, Depends, HTTPException, status
+from fastapi.responses import JSONResponse, HTMLResponse, FileResponse
+from fastapi.templating import Jinja2Templates
+from fastapi.middleware.cors import CORSMiddleware
+from fastapi.security import OAuth2PasswordBearer, OAuth2PasswordRequestForm
+import inspect
+import os
+import posixpath
+import pkg_resources
+import secrets
+from starlette.responses import RedirectResponse
+import traceback
+from typing import List, Optional, Type, TYPE_CHECKING
+import urllib
+import uvicorn
+
+from gradio import utils, queueing
+from gradio.process_examples import load_from_cache, process_example
+
+
+STATIC_TEMPLATE_LIB = pkg_resources.resource_filename("gradio", "templates/")
+STATIC_PATH_LIB = pkg_resources.resource_filename(
+    "gradio", "templates/frontend/static")
+VERSION_FILE = pkg_resources.resource_filename("gradio", "version.txt")
+with open(VERSION_FILE) as version_file:
+    VERSION = version_file.read()
+GRADIO_STATIC_ROOT = "https://gradio.s3-us-west-2.amazonaws.com/{}/static/".format(VERSION)
+
+app = FastAPI()
+app.add_middleware(
+    CORSMiddleware,
+    allow_origins=["*"],
+    allow_methods=["*"],
+    allow_headers=["*"],
+)
+
+templates = Jinja2Templates(directory=STATIC_TEMPLATE_LIB)
+
+###########
+# Auth 
+###########
+
+
+oauth2_scheme = OAuth2PasswordBearer(tokenUrl="/login", auto_error=False)
+
+
+def get_current_user(token: str = Depends(oauth2_scheme)) -> Optional[str]:
+    if token in app.tokens:
+        return app.tokens[token]
+
+
+def is_authenticated(token: str = Depends(oauth2_scheme)) -> bool:
+    return get_current_user(token) is not None     
+
+
+def login_check(is_authenticated: bool = Depends(is_authenticated)):
+    if app.auth is not None and not(is_authenticated):
+        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, 
+                            detail="Not authenticated")
+
+
+@app.get('/token')
+def get_token(token: str = Depends(oauth2_scheme)):
+    return {"token": token}
+
+
+@app.post('/login')
+def login(form_data: OAuth2PasswordRequestForm = Depends()):
+    username, password = form_data.username, form_data.password
+    if ((not callable(app.auth) and username in app.auth 
+        and app.auth[username] == password)
+        or (callable(app.auth) and app.auth.__call__(username, password))):
+        token = secrets.token_urlsafe(16)
+        app.tokens[token] = username
+        return {"access_token": token, "token_type": "bearer"}
```

Comment:
Hello there @abidlabs, was taking a look in this PR,

I would recommend using response model validations instead of bare dictionaries, that way FastAPI would check and filter output data. 
Example: [@app.get("/users/", response_model=List[UserRead])](https://github.com/FarukOzderim/Memory-Vault/blob/151b3fc548540445221d2ff57b0338c6fce313a6/src/listener.py#L55)

Reply:
Hmm interesting, but would that make sense even for this `/login`  endpoint since it's just a token and bearer?

Reply:
Yeah definitely, since it defines the API endpoint documentation, mentioned as [here](https://github.com/gradio-app/gradio/pull/440#issuecomment-1004829267)